Story/vspc 4

namrathaov opened the pull request:
**Summary:**

1. Added a new tab - Exhibition Configuration for staff in path staff/exhibit/econfig.
2. Select a default space for existing/new exhibition.

**Added files:**
DefaultExhibitionRepository.java
IDefaultExhibition.java
DefaultExhibition.java
IDefaultExhibitionManager.java
DefaultExhibitionManager.java
ExhibitionConfigurationController.java
econfig.jsp
econfig_menu.jsp

**Files changed:**
tiles-def.xml
web.xml

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/IDefaultExhibition.java:
```diff
@@ -0,0 +1,14 @@
+
+import edu.asu.diging.vspace.core.model.impl.Space;
+
+public interface IDefaultExhibition extends IVSpaceElement {
```
this should be called IExhibition (there can only be one default, but many exhibitions)

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/DefaultExhibition.java:
```diff
@@ -0,0 +1,65 @@
+
+	@Id
+	@GeneratedValue(generator = "id-generator")
+	@GenericGenerator(name = "id-generator", 	
```
i'm pretty sure you need to give the generator a unique name

namrathaov replied:
[VSPC-4] I have given a unique prefix. I'm not sure about unique name for the generator itself. The module and space have the same generator name.

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/DefaultExhibition.java:
```diff
@@ -0,0 +1,65 @@
+	/* (non-Javadoc)
+	 * @see edu.asu.diging.vspace.core.model.impl.IDefaultExhibition#getId()
+	 */
+	@Override
```
why are there two non-javadoc comments?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/DefaultExhibitionManager.java:
```diff
@@ -0,0 +1,56 @@
+	private DefaultExhibitionRepository exhibitRepo;
+	
+@Override
+public CreationReturnValue storeSpace(DefaultExhibition exhibit) {
```
why is this called 'storeSpace'?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/DefaultExhibitionManager.java:
```diff
@@ -0,0 +1,56 @@
+}
+
+@Override
+public IDefaultExhibition getExhibitionbyId(String id){
```
'by' needs to be capitalized

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,52 @@
+	private DefaultExhibitionRepository exhibitRepo;
+	
+	@RequestMapping("/staff/exhibit/econfig")
+	public String listSpaces(Model model) {
```
why is this called 'listSpaces'?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,52 @@
+	}
+	
+	@RequestMapping(value = "/staff/exhibit/config_add", method = RequestMethod.POST)
+	public String addDefaultSpace(@RequestParam(required=false,name="dexhibit") String exhibitID, @RequestParam("dspace") String spaceID) throws IOException {
```
why is this called 'addDefaultSpace'?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,52 @@
+		return "staff/exhibit/econfig";
+	}
+	
+	@RequestMapping(value = "/staff/exhibit/config_add", method = RequestMethod.POST)
```
bad url path, replace underscore with slash

jdamerow replied:
or even better just remove '_add'

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,52 @@
+	@Autowired
+	private DefaultExhibitionRepository exhibitRepo;
+	
+	@RequestMapping("/staff/exhibit/econfig")
```
since you use /staff/exhibit/config/add below, this should be /staff/exhibit/config

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,52 @@
+		DefaultExhibition exhibit;
+		Space space = (Space) spaceManager.getSpace(spaceID);
+		if(exhibitID.equals("New"))
+			exhibit = new DefaultExhibition();
```
that's what the factory is for

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,52 @@
+	@RequestMapping("/staff/exhibit/econfig")
+	public String listSpaces(Model model) {
+		if(exhibitRepo.count()>0) {
+		model.addAttribute("exhibit", exhibitRepo.findAll());
```
what does 'findAll' return when there are no exhibitions? can't you just add it as attribute always.

jdamerow commented on vspace/src/main/webapp/WEB-INF/web.xml:
```diff
@@ -37,21 +39,34 @@
+	<filter-mapping>
+		<filter-name>SpringOpenEntityManagerInViewFilter</filter-name>
+		<url-pattern>/*</url-pattern>
+	</filter-mapping>
```
what does this do?

namrathaov replied:
[VSPC-4] I get this error when I try to save an Exhibition: 
org.hibernate.LazyInitializationException: failed to lazily initialize a collection of role: edu.asu.diging.vspace.core.model.impl.Space.spaceLinks, could not initialize proxy - no Session

Since the Exhibition is getting the space and the space links are lazy loaded in the SpaceManager along with it, this is happening. Is there a better way to resolve this issue?

jdamerow replied:
No, this is actually pretty cool. I didn't know this existed. Do you understand what the issue was and how you solved it?

namrathaov replied:
Yes. I worked on lazy loading a bit before. But now I clearly understood. I read an article that explained what causes this error and how lazy loading is used. 

One article said using this open entity manager may setback performance depending on how it's used.

namrathaov commented:
Files Modified in Reopened PR: 
Renamed IDefaultExhibition.java -> IExhibition.java
Renamed DefaultExhibition.java -> Exhibition.java
Renamed DefaultExhibitionManager.java -> ExhibitionManager.java
Renamed econfig.jsp -> config.jsp
Renamed econfig_menu -> config_menu.jsp
ExhibitionConfigurationController.java

Path changed to space/exhibit/config for both get and post.

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/factory/impl/ExhibitionFactory.java:
```diff
@@ -0,0 +1,23 @@
+		else
+			return exhibitManager.getExhibitionById(exhibitID);
+	}
+}
```
pleaes follow our code style guide

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/factory/impl/ExhibitionFactory.java:
```diff
@@ -0,0 +1,23 @@
+	private ExhibitionManager exhibitManager;
+	@Override
+	public IExhibition createSpace(String exhibitID) {
+		if(exhibitID.equals("New"))
```
the hard-coded string already gives you a hint that this is not an ideal solution. Couple of problems:
- if you need to decide by a passed id if something should be create or not, use null not a "random" string
- this method is a factory method for creating spaces, so all it should do is creating spaces (no retrieval of spaces form the db)
- since your manager class will most likely use the factory at some point, you have a circular dependency here, which is a big no no

namrathaov replied:
For #1, I have used empty string, I could not handle null pointer exception while creating Exhibition in ExhibitionManager.

jdamerow replied:
what do you mean by that?

namrathaov replied:
For the New exhibition dropdown option, I cannot set the option value to null because it throws error even before I can handle it in my backend. So I have set the option value to empty string instead of a random string.

jdamerow replied:
but this is in the exhibition factory, you can call that method with null, can't you?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Exhibition.java:
```diff
@@ -0,0 +1,49 @@
+import edu.asu.diging.vspace.core.model.IExhibition;
+
+@Entity
+public class Exhibition extends VSpaceElement implements IExhibition {
```
needs javadoc

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/model/impl/Exhibition.java:
```diff
@@ -0,0 +1,49 @@
+
+	@Id
+	@GeneratedValue(generator = "id-generator")
+	@GenericGenerator(name = "id-generator", 	
```
you can choose the generator id more or less as you please as long as it is unique (that there are two generators with the same name already doesn't mean it's right ;)

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ExhibitionManager.java:
```diff
@@ -0,0 +1,56 @@
+	exhibitRepo.findAll().forEach(s -> defaultExhibition.add(s));
+	return defaultExhibition;
+	}
+}
```
formatting!

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ExhibitionManager.java:
```diff
@@ -0,0 +1,56 @@
+@Transactional
+@Service
+@PropertySource("classpath:/config.properties")
+public class ExhibitionManager implements IExhibitionManager {
```
needs javadoc and test cases

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/factory/impl/ExhibitionFactory.java:
```diff
@@ -0,0 +1,23 @@
+import edu.asu.diging.vspace.core.services.impl.ExhibitionManager;
+
+@Service
+public class ExhibitionFactory implements IExhibitionFactory {
```
needs test cases

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ExhibitionManager.java:
```diff
@@ -0,0 +1,56 @@
+
+@Override
+public List<IExhibition> getAllExhibitions() {
+	List<IExhibition> defaultExhibition = new ArrayList<>();
```
why is this list called 'defaultExhibition'?

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ExhibitionManager.java:
```diff
@@ -0,0 +1,56 @@
+@Override
+public IExhibition getExhibitionById(String id){
+	
+	Optional<Exhibition> defaultExhibition = exhibitRepo.findById(id);
```
variable should not be called defaultExhibition unless it's a default exhibition (which it is not)

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,49 @@
+	private ExhibitionRepository exhibitRepo;
+	@Autowired
+	private ExhibitionFactory exhibitFactory;
+	@RequestMapping("/staff/exhibit/config")
```
there should be an empty line above this line

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,49 @@
+	
+	@RequestMapping(value = "/staff/exhibit/config", method = RequestMethod.POST)
+	public String createOrUpdateExhibition(@RequestParam(required=false,name="dexhibit") String exhibitID, @RequestParam("dspace") String spaceID) throws IOException {
+		Exhibition exhibit;
```
put where this is needed

jdamerow commented on vspace/src/main/java/edu/asu/diging/vspace/web/staff/ExhibitionConfigurationController.java:
```diff
@@ -0,0 +1,49 @@
+	@RequestMapping(value = "/staff/exhibit/config", method = RequestMethod.POST)
+	public String createOrUpdateExhibition(@RequestParam(required=false,name="dexhibit") String exhibitID, @RequestParam("dspace") String spaceID) throws IOException {
+		Exhibition exhibit;
+		Space space = (Space) spaceManager.getSpace(spaceID);
```
why are you casting this to a space? should use the interface here.
